File: url_shortening_service_flask/controller/url_controller.py
from flask import Blueprint, request, jsonify, abort
from service.url_service import shorten_url, get_original_url

url_controller = Blueprint("url_controller", __name__)

@url_controller.route("/shorten", methods=["POST"])
def create_short_url():
    # Check for `url` in query parameters or in the JSON body
    original_url = request.args.get("url", None)
    data = request.get_json(silent=True)  # Avoid error if no JSON body
    body_url = data.get("url") if data else None

    # Validate: Only one of `url` query param or `url` in body should be present
    if original_url and body_url:
        abort(400, description="Provide 'url' as either a query parameter or in the request body, but not both.")
    elif not original_url and not body_url:
        abort(400, description="URL is required as a query parameter or in the request body.")

    # Use whichever `url` is provided
    url_to_shorten = original_url if original_url else body_url
    short_url = shorten_url(url_to_shorten)

    return jsonify({"short_url": short_url, "original_url": url_to_shorten})

@url_controller.route("/<short_id>", methods=["GET"])
def retrieve_original_url(short_id):
    original_url = get_original_url(short_id)
    if original_url is None:
        abort(404, description="URL not found")

    return jsonify({"original_url": original_url})

File: url-shortening-service/src/main/java/com/project/url_shortening_service/controller/UrlController.java
package com.project.url_shortening_service.controller;

import com.project.url_shortening_service.dto.request.ShortenUrlRequest;
import com.project.url_shortening_service.dto.response.ShortenUrlResponse;
import com.project.url_shortening_service.exception.InvalidUrlInputException;
import com.project.url_shortening_service.model.UrlMapping;
import com.project.url_shortening_service.service.UrlService;
import com.project.url_shortening_service.validator.ValidHttpLink;
import jakarta.validation.Valid;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/url")
@Slf4j
@Validated
public class UrlController {

    private final UrlService urlService;

    @Autowired
    public UrlController(UrlService urlService) {
        this.urlService = urlService;
    }

    @PostMapping("/shorten")
    public ResponseEntity<String> shortenUrl(
            @RequestParam(value = "url", required = false) String originalUrl,
            @RequestBody(required = false) @Valid ShortenUrlRequest shortenUrlRequest) {

        log.info("in POST/user/shorten");
        log.info("Incoming: " + "originalUrl="+originalUrl + " " +
                "shortenUrlRequest:" + shortenUrlRequest );

        // Check that either originalUrl or shortenUrlRequest is provided
        if ((originalUrl == null && shortenUrlRequest == null) ||
                (originalUrl != null && shortenUrlRequest != null)) {
            throw new InvalidUrlInputException("Provide either 'url' as a query parameter or " +
                    "in the request body, but not both.");
        }

        // Determine which URL to shorten
        String urlToShorten = originalUrl != null ? originalUrl : shortenUrlRequest.getUrl();
        log.info("urlToShorten: " + urlToShorten);
        // Validate the final URL to shorten
        if (originalUrl == null) {
            if (shortenUrlRequest == null || shortenUrlRequest.getUrl() == null) {
                throw new InvalidUrlInputException("URL to shorten must not be null.");
            }
            urlToShorten = shortenUrlRequest.getUrl();
        }

        String response = urlService.shortenUrl(urlToShorten);
        log.info("Response: " + response);
        return ResponseEntity.ok(response);
    }

    @GetMapping("/{shorten_url}")
    public ResponseEntity<String> getActualUrl(@PathVariable("shorten_url") String shortUrl) {
        String originalUrl = urlService.getActualUrl(shortUrl);
        return ResponseEntity.ok(originalUrl);
    }

    @GetMapping("/{shorten_url}/all")
    public ResponseEntity<ShortenUrlResponse> getAllMapping(@PathVariable("shorten_url") String shortUrl) {
        UrlMapping mapping = urlService.getMappingData(shortUrl);
        ShortenUrlResponse response = new ShortenUrlResponse(mapping.getId(), mapping.getUrl(), mapping.getActualUrl());
        return ResponseEntity.ok(response);
    }
}
